File: dev-datang2/src/main/java/com/example/controller/RecallHybrid.java
package com.example.controller;


import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.elasticsearch._types.query_dsl.*;
import co.elastic.clients.elasticsearch.core.SearchRequest;
import co.elastic.clients.elasticsearch.core.SearchResponse;
import co.elastic.clients.elasticsearch.core.search.Hit;
import co.elastic.clients.json.JsonData;
import co.elastic.clients.json.jackson.JacksonJsonpMapper;
import co.elastic.clients.transport.rest_client.RestClientTransport;
import org.apache.http.HttpHost;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.elasticsearch.client.RestClient;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.stream.Collectors;

public class RecallHybrid {

    private static final String ES_URL = "http://localhost:9200";
    private static final String INDEX_NAME = "datang_docs_index1";

    private static final String EMBEDDING_API_URL = "http://10.3.24.46:9997/v1/embeddings";
    private static final String EMBEDDING_API_KEY = "123";

    private static final Gson gson = new Gson();

    private static ElasticsearchClient esClient;

    static {
        RestClient restClient = RestClient.builder(HttpHost.create(ES_URL)).build();
        RestClientTransport transport = new RestClientTransport(restClient, new JacksonJsonpMapper());
        esClient = new ElasticsearchClient(transport);
    }

    // 调用 Embedding API
    public static List<Float> getEmbedding(String text) throws Exception {
        try (CloseableHttpClient client = HttpClients.createDefault()) {
            HttpPost post = new HttpPost(EMBEDDING_API_URL);
            post.setHeader("Content-Type", "application/json");
            post.setHeader("Authorization", "Bearer " + EMBEDDING_API_KEY);

            JsonObject body = new JsonObject();
            body.addProperty("model", "bge-large-zh-v1.5");
            body.add("input", gson.toJsonTree(text));

            post.setEntity(new StringEntity(body.toString(), StandardCharsets.UTF_8));

            try (CloseableHttpResponse response = client.execute(post)) {
                BufferedReader reader = new BufferedReader(
                        new InputStreamReader(response.getEntity().getContent(), StandardCharsets.UTF_8));
                String result = reader.lines().collect(Collectors.joining());
                JsonObject json = gson.fromJson(result, JsonObject.class);

                JsonArray arr = json.getAsJsonArray("data")
                        .get(0).getAsJsonObject()
                        .getAsJsonArray("embedding");

                List<Float> embedding = new ArrayList<>();
                arr.forEach(e -> embedding.add(e.getAsFloat()));
                System.out.println("嵌入的查询文本: " + text);
                return embedding;
            }
        }
    }

    // 混合搜索
    public static List<Map<String, Object>> hybridSearchIndicator(
            String queryText, List<Float> queryVector, String companyName) throws Exception {

        // ===== 1. 关键词检索 =====
        Query keywordQuery = BoolQuery.of(b -> b
                .must(TermQuery.of(t -> t.field("company_name.keyword").value(companyName))._toQuery())
                .must(MatchQuery.of(m -> m.field("content").query(queryText).analyzer("ik_max_word"))._toQuery())
        )._toQuery();

        SearchResponse<Map> keywordRes = esClient.search(s -> s
                .index(INDEX_NAME)
                .query(keywordQuery)
                .size(10), Map.class);

        // ===== 2. 向量检索 =====
        Map<String, JsonData> params = new HashMap<>();
        params.put("query_vector", JsonData.of(queryVector));

        Query vectorQuery = ScriptScoreQuery.of(s -> s
                .query(BoolQuery.of(b -> b
                        .must(TermQuery.of(t -> t.field("company_name.keyword").value(companyName))._toQuery())
                        .must(ExistsQuery.of(e -> e.field("content"))._toQuery())
                )._toQuery())
                .script(script -> script
                        .inline(inline -> inline
                                .source("cosineSimilarity(params.query_vector, 'embedding') + 1.0")
                                .params(params)
                        )
                )
        )._toQuery();

        SearchResponse<Map> vectorRes = esClient.search(s -> s
                .index(INDEX_NAME)
                .query(vectorQuery)
                .size(10), Map.class);

        // ===== 3. 合并逻辑 =====
        Map<String, Map<String, Object>> merged = new HashMap<>();

        for (Hit<Map> hit : keywordRes.hits().hits()) {
            Map<String, Object> src = hit.source();
            String key = src.get("company_name") + "|" + src.get("dataset_id") + "|" + src.get("content");
            Map<String, Object> val = new HashMap<>();
            // 获取最大数据

            val.put("source", src);
            val.put("bm25_score", hit.score() != null ? hit.score() : 0.0);
            val.put("vector_score", 0.0);
            merged.put(key, val);
        }

        for (Hit<Map> hit : vectorRes.hits().hits()) {
            Map<String, Object> src = hit.source();
            String key = src.get("company_name") + "|" + src.get("dataset_id") + "|" + src.get("content");
            Map<String, Object> val = merged.getOrDefault(key, new HashMap<>());
            if (!val.containsKey("source")) val.put("source", src);
            val.put("bm25_score", val.getOrDefault("bm25_score", 0.0));
            val.put("vector_score", hit.score() != null ? hit.score() : 0.0);
            merged.put(key, val);
        }

        // 计算新得分
        List<Map<String, Object>> finalResults = new ArrayList<>();
        for (Map<String, Object> val : merged.values()) {
            double bm25 = (double) val.get("bm25_score");
            double vec = (double) val.get("vector_score");
            double score;
            if (bm25 > 0 && vec > 0) {
                score = bm25 * 0.5 + vec * 0.5;
            } else {
                score = Math.max(bm25, vec) * 0.5;
            }
            Map<String, Object> src = (Map<String, Object>) val.get("source");
            Map<String, Object> item = new HashMap<>();
            item.put("score", score);
            item.put("company_name", src.get("company_name"));
            item.put("dataset_id", src.get("dataset_id"));
            item.put("content", src.get("content"));
            finalResults.add(item);
        }

        finalResults.sort((a, b) -> Double.compare((double) b.get("score"), (double) a.get("score")));
        System.out.println("关键词+向量召回的指标数据总数:" + finalResults.size());
        return finalResults.stream().limit(10).collect(Collectors.toList());
    }

    // 过滤高分数据
    public static List<Map<String, Object>> getHighScore(List<Map<String, Object>> results) {
        List<Map<String, Object>> high = new ArrayList<>();
        if (results.isEmpty()) return high;
        double score1 = (double) results.get(0).get("score");
        for (Map<String, Object> item : results) {
            double score = (double) item.get("score");
            if ((score1 - score) / score1 <= 0.3) {
                high.add(item);
            }
        }
        return high;
    }

    // 归一化数据为0~1之间
    public static Float normalizeScore(double score) {


        return (float) (score - 0.5) * 2;
    }

    // 合并相同公司内容
    public static Map<String, List<Map<String, List<String>>>> sameCompanyContentMerged(
            List<Map<String, Object>> data) {

        Map<String, List<Map<String, List<String>>>> results = new HashMap<>();

        for (Map<String, Object> d : data) {
            String company = (String) d.get("company_name");
            String datasetId = (String) d.get("dataset_id");
            String content = (String) d.get("content");

            results.putIfAbsent(company, new ArrayList<>());

            boolean found = false;
            for (Map<String, List<String>> item : results.get(company)) {
                if (item.containsKey(datasetId)) {
                    item.get(datasetId).add(content);
                    found = true;
                    break;
                }
            }
            if (!found) {
                Map<String, List<String>> newEntry = new HashMap<>();
                newEntry.put(datasetId, new ArrayList<>(Collections.singletonList(content)));
                results.get(company).add(newEntry);
            }
        }
        return results;
    }

    public static void main(String[] args) throws Exception {
        String queryText = "公司名称:大唐江苏有限公司 时间:六月份 指标名称:燃机汇总上网电量月指标（大屏）是多少";

        List<Float> queryVector = getEmbedding(queryText);

        List<Map<String, Object>> results = hybridSearchIndicator(queryText, queryVector, "大唐江苏有限公司");

        List<Map<String, Object>> highIndicatorRes = getHighScore(results);

        Map<String, List<Map<String, List<String>>>> mergedRes = sameCompanyContentMerged(highIndicatorRes);

        System.out.println("查询的指标数据:\n" + gson.toJson(mergedRes));

        for (Map.Entry<String, List<Map<String, List<String>>>> entry : mergedRes.entrySet()) {
            String company = entry.getKey();
            for (Map<String, List<String>> item : entry.getValue()) {
                for (Map.Entry<String, List<String>> dt : item.entrySet()) {
                    System.out.println("公司名称: " + company + " - 数据集ID:" + dt.getKey() + " - 对应的指标名称:" + dt.getValue());
                }
            }
        }
    }
}

File: indicator_data/recall_hybrid.py
import os
import openai
from dotenv import load_dotenv
from elasticsearch import Elasticsearch

es = Elasticsearch("http://localhost:9200", verify_certs=False)
index_name = "datang_docs_index1"
load_dotenv()
os.environ["OPENAI_BASE_URL"] = "http://10.3.24.46:9997/v1"
os.environ["OPENAI_API_KEY"] = "123"

print(es.indices.get_mapping(index=index_name))

# 0.初始化嵌入模型对象
client = openai.Client()


def get_embedding(text):
    res = client.embeddings.create(input=text, model="bge-large-zh-v1.5")
    print("嵌入的查询文本:", text)
    return res.data[0].embedding


def hybrid_search_indicator(query_text, query_vector, company_name):
    # ====== 1. 关键词检索 ======
    keyword_query = {
        "bool": {
            "must": [
                {"term": {"company_name.keyword": company_name}},  # 限定公司名称
            ],
            "should": [
                {"match": {
                    "content": {
                        "query": query_text,
                        "analyzer": "ik_max_word"
                    }
                }}
            ]
        }
    }
    keyword_res = es.search(index=index_name, query=keyword_query, size=10)["hits"]["hits"]

    # ====== 2. 向量检索 ======
    vector_query = {
        "script_score": {
            "query": {
                "bool": {
                    "must": [
                        {"term": {"company_name.keyword": company_name}},  # 精准匹配公司名
                        {"exists": {"field": "content"}}
                    ]
                }
            },
            "script": {
                "source": "(cosineSimilarity(params.query_vector, 'embedding') + 1.0)/2",
                "params": {"query_vector": query_vector}
            }
        }
    }

    vector_res = es.search(index=index_name, query=vector_query, size=10)["hits"]["hits"]

    # ====== 3. 合并逻辑 ======
    merged = {}
    # 先存放关键词检索结果
    for hit in keyword_res:
        key = (hit["_source"]["company_name"], hit["_source"]["dataset_id"], hit["_source"]["content"])
        merged[key] = {
            "source": hit["_source"],
            "bm25_score": hit["_score"],
            "vector_score": 0.0  # 暂无
        }

    # 再存放向量检索结果
    for hit in vector_res:
        key = (hit["_source"]["company_name"], hit["_source"]["dataset_id"], hit["_source"]["content"])
        if key in merged:
            merged[key]["vector_score"] = hit["_score"]
        else:
            merged[key] = {
                "source": hit["_source"],
                "bm25_score": 0.0,
                "vector_score": hit["_score"]
            }

    # 计算新得分
    final_results = []
    for key, val in merged.items():
        bm25 = val["bm25_score"]
        vec = val["vector_score"]

        if bm25 > 0 and vec > 0:  # 同时命中
            score = bm25 * 0.5 + vec * 0.5
        else:  # 只命中一个
            score = max(bm25, vec) * 0.5

        final_results.append({
            "score": score,
            "company_name": val["source"]["company_name"],
            "dataset_id": val["source"]["dataset_id"],
            "content": val["source"]["content"]
        })

    # 排序
    final_results.sort(key=lambda x: x["score"], reverse=True)
    print("关键词+向量召回的指标数据总数:", len(final_results))
    # 取前10个
    top_10_results = final_results[:10]
    return top_10_results


# 用户输入问题
# query_text = "大唐集团南京分公司的滨海公司年完成率（小科）是多少？"
query_text = "公司名称:大唐江苏有限公司 时间:六月份 指标名称:泰州热电月度供电煤耗,吕四港公司年度供电煤耗 这两个指标的值是多少"
# query_text = "公司名称:大唐江苏有限公司 时间:六月份 指标名称:淮安光伏场站发电收益日指标（大屏）6月份的平均值?"

query_vector = get_embedding(query_text)

# 检索ES数据
results = hybrid_search_indicator(query_text, query_vector, company_name="大唐江苏有限公司")


# 过滤ES返回数据中的低分数据
def get_high_score(results_list):
    # 获取高分行召回数据
    record_high_score = []
    # 获取第一个得分最高最相关的数据
    score1 = results_list[0]["score"]
    for index, item in enumerate(results_list):
        score_x = results_list[index]["score"]
        if (score1 - score_x) / score1 <= 0.5:
            record_high_score.append(item)
    return record_high_score


high_indicator_res = get_high_score(results)

need_merged_company = []
for item in high_indicator_res:
    need_merged_company.append({
        item["company_name"]: {"score": item["score"],
                               "dataset_id": item["dataset_id"],
                               "content": item["content"]}})


# 组装成符合要求的数据
def same_company_content_merged(data: list[dict]) -> dict:
    results = {}
    for d in data:
        for company, v in d.items():
            dataset_id = v["dataset_id"]
            content = v["content"]

            if company not in results:
                results[company] = []

            # 查找是否已有该 dataset_id
            found = False
            for item in results[company]:
                if dataset_id in item:
                    item[dataset_id].append(content)
                    found = True
                    break

            if not found:
                results[company].append({dataset_id: [content]})

    return results


merged_res = same_company_content_merged(need_merged_company)

print("查询的指标数据:\n", merged_res, "\n")

# 打印结果
for k, v in merged_res.items():
    for each_item in v:
        for dt_id, dt_indicator in each_item.items():
            print(f" 公司名称: {k} - 数据集ID:{dt_id} - 对应的指标名称: {dt_indicator}")
